File: integrations/langchain-ts/src/ephemeral.ts
/**
 * Xache Ephemeral Context for LangChain.js
 * Create, manage, and promote short-lived working memory sessions
 */

import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import {
  XacheClient,
  DID,
  type XacheSigner,
  type XacheWalletProvider,
} from '@xache/sdk';

// =============================================================================
// Configuration
// =============================================================================

export interface EphemeralToolConfig {
  /** Wallet address for authentication */
  walletAddress: string;
  /** Private key for signing (optional if signer or walletProvider is provided) */
  privateKey?: string;
  /** External signer (alternative to privateKey) */
  signer?: XacheSigner;
  /** Wallet provider for lazy signer resolution */
  walletProvider?: XacheWalletProvider;
  /** Encryption key for use with external signers */
  encryptionKey?: string;
  /** API URL (defaults to https://api.xache.xyz) */
  apiUrl?: string;
  /** Chain: 'base' or 'solana' */
  chain?: 'base' | 'solana';
}

function createClient(config: EphemeralToolConfig): XacheClient {
  const chainPrefix = config.chain === 'solana' ? 'sol' : 'evm';
  const did = `did:agent:${chainPrefix}:${config.walletAddress.toLowerCase()}` as DID;

  return new XacheClient({
    apiUrl: config.apiUrl || 'https://api.xache.xyz',
    did,
    privateKey: config.privateKey,
    signer: config.signer,
    walletProvider: config.walletProvider,
    encryptionKey: config.encryptionKey,
  });
}

// =============================================================================
// Factory Functions
// =============================================================================

/**
 * Create a tool for creating an ephemeral working memory session.
 */
export function createEphemeralCreateSessionTool(
  config: EphemeralToolConfig
): DynamicStructuredTool {
  const client = createClient(config);

  return new DynamicStructuredTool({
    name: 'xache_ephemeral_create_session',
    description:
      'Create a new ephemeral working memory session. ' +
      'Returns a session key for storing temporary data in slots.',
    schema: z.object({
      ttlSeconds: z.number().optional().describe('Session time-to-live in seconds (default: 3600)'),
      maxWindows: z.number().optional().describe('Maximum renewal windows (default: 5)'),
    }),
    func: async ({ ttlSeconds, maxWindows }) => {
      const session = await client.ephemeral.createSession({
        ttlSeconds,
        maxWindows,
      });

      return [
        `Created ephemeral session.`,
        `Session Key: ${session.sessionKey}`,
        `Status: ${session.status}`,
        `TTL: ${session.ttlSeconds}s`,
        `Expires: ${session.expiresAt}`,
      ].join('\n');
    },
  });
}

/**
 * Create a tool for writing data to an ephemeral slot.
 */
export function createEphemeralWriteSlotTool(
  config: EphemeralToolConfig
): DynamicStructuredTool {
  const client = createClient(config);

  return new DynamicStructuredTool({
    name: 'xache_ephemeral_write_slot',
    description:
      'Write data to an ephemeral session slot. ' +
      'Slots: conversation, facts, tasks, cache, scratch, handoff.',
    schema: z.object({
      sessionKey: z.string().describe('The ephemeral session key'),
      slot: z.enum(['conversation', 'facts', 'tasks', 'cache', 'scratch', 'handoff']).describe('Slot name'),
      data: z.record(z.string(), z.unknown()).describe('Data to write to the slot'),
    }),
    func: async ({ sessionKey, slot, data }) => {
      await client.ephemeral.writeSlot(sessionKey, slot as any, data as Record<string, unknown>);
      return `Wrote data to slot "${slot}" in session ${sessionKey.substring(0, 12)}...`;
    },
  });
}

/**
 * Create a tool for reading data from an ephemeral slot.
 */
export function createEphemeralReadSlotTool(
  config: EphemeralToolConfig
): DynamicStructuredTool {
  const client = createClient(config);

  return new DynamicStructuredTool({
    name: 'xache_ephemeral_read_slot',
    description:
      'Read data from an ephemeral session slot. ' +
      'Slots: conversation, facts, tasks, cache, scratch, handoff.',
    schema: z.object({
      sessionKey: z.string().describe('The ephemeral session key'),
      slot: z.enum(['conversation', 'facts', 'tasks', 'cache', 'scratch', 'handoff']).describe('Slot name'),
    }),
    func: async ({ sessionKey, slot }) => {
      const data = await client.ephemeral.readSlot(sessionKey, slot as any);
      return JSON.stringify(data, null, 2);
    },
  });
}

/**
 * Create a tool for promoting an ephemeral session to persistent memory.
 */
export function createEphemeralPromoteTool(
  config: EphemeralToolConfig
): DynamicStructuredTool {
  const client = createClient(config);

  return new DynamicStructuredTool({
    name: 'xache_ephemeral_promote',
    description:
      'Promote an ephemeral session to persistent memory. ' +
      'Extracts valuable data from slots and stores as permanent memories.',
    schema: z.object({
      sessionKey: z.string().describe('The ephemeral session key to promote'),
    }),
    func: async ({ sessionKey }) => {
      const result = await client.ephemeral.promoteSession(sessionKey);

      let output = `Promoted session ${sessionKey.substring(0, 12)}...\n`;
      output += `Memories created: ${result.memoriesCreated}\n`;
      if (result.memoryIds.length > 0) {
        output += `Memory IDs: ${result.memoryIds.join(', ')}\n`;
      }
      if (result.receiptId) {
        output += `Receipt: ${result.receiptId}`;
      }
      return output;
    },
  });
}

/**
 * Create a tool for getting ephemeral session status and details.
 */
export function createEphemeralStatusTool(
  config: EphemeralToolConfig
): DynamicStructuredTool {
  const client = createClient(config);

  return new DynamicStructuredTool({
    name: 'xache_ephemeral_status',
    description:
      'Get the status and details of an ephemeral session. ' +
      'Shows active slots, size, TTL, and window information.',
    schema: z.object({
      sessionKey: z.string().describe('The ephemeral session key'),
    }),
    func: async ({ sessionKey }) => {
      const session = await client.ephemeral.getSession(sessionKey);

      if (!session) {
        return `Session ${sessionKey.substring(0, 12)}... not found.`;
      }

      return [
        `Session: ${session.sessionKey.substring(0, 12)}...`,
        `Status: ${session.status}`,
        `Window: ${session.window}/${session.maxWindows}`,
        `TTL: ${session.ttlSeconds}s`,
        `Expires: ${session.expiresAt}`,
        `Active Slots: ${session.activeSlots.length > 0 ? session.activeSlots.join(', ') : 'none'}`,
        `Total Size: ${session.totalSize} bytes`,
        `Cumulative Cost: $${session.cumulativeCost.toFixed(4)}`,
      ].join('\n');
    },
  });
}

// =============================================================================
// Class Wrappers
// =============================================================================

export class XacheEphemeralCreateSessionTool {
  private tool: DynamicStructuredTool;

  constructor(config: EphemeralToolConfig) {
    this.tool = createEphemeralCreateSessionTool(config);
  }

  asTool(): DynamicStructuredTool {
    return this.tool;
  }
}

export class XacheEphemeralWriteSlotTool {
  private tool: DynamicStructuredTool;

  constructor(config: EphemeralToolConfig) {
    this.tool = createEphemeralWriteSlotTool(config);
  }

  asTool(): DynamicStructuredTool {
    return this.tool;
  }
}

export class XacheEphemeralReadSlotTool {
  private tool: DynamicStructuredTool;

  constructor(config: EphemeralToolConfig) {
    this.tool = createEphemeralReadSlotTool(config);
  }

  asTool(): DynamicStructuredTool {
    return this.tool;
  }
}

export class XacheEphemeralPromoteTool {
  private tool: DynamicStructuredTool;

  constructor(config: EphemeralToolConfig) {
    this.tool = createEphemeralPromoteTool(config);
  }

  asTool(): DynamicStructuredTool {
    return this.tool;
  }
}

export class XacheEphemeralStatusTool {
  private tool: DynamicStructuredTool;

  constructor(config: EphemeralToolConfig) {
    this.tool = createEphemeralStatusTool(config);
  }

  asTool(): DynamicStructuredTool {
    return this.tool;
  }
}

File: integrations/langchain-python/xache_langchain/ephemeral.py
"""
Xache Ephemeral Context for LangChain
Create, manage, and promote short-lived working memory sessions
"""

from typing import Any, Dict, List, Optional
from langchain.tools import BaseTool
from pydantic import BaseModel, Field

from xache import XacheClient


# =============================================================================
# Input Schemas
# =============================================================================


class EphemeralCreateSessionInput(BaseModel):
    """Input for ephemeral create session tool"""

    ttl_seconds: int = Field(default=3600, description="Session TTL in seconds (default: 3600)")
    max_windows: int = Field(default=5, description="Maximum renewal windows (default: 5)")


class EphemeralWriteSlotInput(BaseModel):
    """Input for ephemeral write slot tool"""

    session_key: str = Field(description="The ephemeral session key")
    slot: str = Field(
        description="Slot name (conversation, facts, tasks, cache, scratch, handoff)"
    )
    data: Dict[str, Any] = Field(description="Data to write to the slot")


class EphemeralReadSlotInput(BaseModel):
    """Input for ephemeral read slot tool"""

    session_key: str = Field(description="The ephemeral session key")
    slot: str = Field(
        description="Slot name (conversation, facts, tasks, cache, scratch, handoff)"
    )


class EphemeralPromoteInput(BaseModel):
    """Input for ephemeral promote tool"""

    session_key: str = Field(description="The ephemeral session key to promote")


class EphemeralStatusInput(BaseModel):
    """Input for ephemeral status tool"""

    session_key: str = Field(description="The ephemeral session key")


# =============================================================================
# Sync helper
# =============================================================================


def _run_sync(coro: Any) -> Any:
    """Run an async coroutine synchronously (handles nested loops)."""
    import asyncio

    try:
        loop = asyncio.get_event_loop()
        if loop.is_running():
            import concurrent.futures

            with concurrent.futures.ThreadPoolExecutor() as pool:
                return pool.submit(asyncio.run, coro).result()
        else:
            return loop.run_until_complete(coro)
    except RuntimeError:
        return asyncio.run(coro)


# =============================================================================
# Tools
# =============================================================================


class XacheEphemeralCreateSessionTool(BaseTool):
    """Create a new ephemeral working memory session."""

    name: str = "xache_ephemeral_create_session"
    description: str = (
        "Create a new ephemeral working memory session. "
        "Returns a session key for storing temporary data in slots."
    )
    args_schema: type[BaseModel] = EphemeralCreateSessionInput

    wallet_address: str
    private_key: Optional[str] = Field(default=None, exclude=True)
    signer: Optional[Any] = Field(default=None, exclude=True)
    wallet_provider: Optional[Any] = Field(default=None, exclude=True)
    encryption_key: Optional[str] = Field(default=None, exclude=True)
    api_url: str = "https://api.xache.xyz"
    chain: str = "base"

    _client: Optional[XacheClient] = None

    def __init__(self, **kwargs: Any) -> None:
        import os

        if "api_url" not in kwargs or kwargs["api_url"] is None:
            kwargs["api_url"] = os.environ.get("XACHE_API_URL", "https://api.xache.xyz")
        super().__init__(**kwargs)
        chain_prefix = "sol" if self.chain == "solana" else "evm"
        did = f"did:agent:{chain_prefix}:{self.wallet_address.lower()}"
        self._client = XacheClient(
            api_url=self.api_url,
            did=did,
            private_key=self.private_key,
            signer=self.signer,
            wallet_provider=self.wallet_provider,
            encryption_key=self.encryption_key,
        )

    def _run(self, ttl_seconds: int = 3600, max_windows: int = 5) -> str:
        async def _create() -> Any:
            async with self._client as client:
                return await client.ephemeral.create_session(
                    ttl_seconds=ttl_seconds, max_windows=max_windows,
                )

        session = _run_sync(_create())
        return (
            f"Created ephemeral session.\n"
            f"Session Key: {session.session_key}\n"
            f"Status: {session.status}\n"
            f"TTL: {session.ttl_seconds}s\n"
            f"Expires: {session.expires_at}"
        )

    async def _arun(self, ttl_seconds: int = 3600, max_windows: int = 5) -> str:
        async with self._client as client:
            session = await client.ephemeral.create_session(
                ttl_seconds=ttl_seconds, max_windows=max_windows,
            )
        return (
            f"Created ephemeral session.\n"
            f"Session Key: {session.session_key}\n"
            f"Status: {session.status}\n"
            f"TTL: {session.ttl_seconds}s\n"
            f"Expires: {session.expires_at}"
        )


class XacheEphemeralWriteSlotTool(BaseTool):
    """Write data to an ephemeral session slot."""

    name: str = "xache_ephemeral_write_slot"
    description: str = (
        "Write data to an ephemeral session slot. "
        "Slots: conversation, facts, tasks, cache, scratch, handoff."
    )
    args_schema: type[BaseModel] = EphemeralWriteSlotInput

    wallet_address: str
    private_key: Optional[str] = Field(default=None, exclude=True)
    signer: Optional[Any] = Field(default=None, exclude=True)
    wallet_provider: Optional[Any] = Field(default=None, exclude=True)
    encryption_key: Optional[str] = Field(default=None, exclude=True)
    api_url: str = "https://api.xache.xyz"
    chain: str = "base"

    _client: Optional[XacheClient] = None

    def __init__(self, **kwargs: Any) -> None:
        import os

        if "api_url" not in kwargs or kwargs["api_url"] is None:
            kwargs["api_url"] = os.environ.get("XACHE_API_URL", "https://api.xache.xyz")
        super().__init__(**kwargs)
        chain_prefix = "sol" if self.chain == "solana" else "evm"
        did = f"did:agent:{chain_prefix}:{self.wallet_address.lower()}"
        self._client = XacheClient(
            api_url=self.api_url,
            did=did,
            private_key=self.private_key,
            signer=self.signer,
            wallet_provider=self.wallet_provider,
            encryption_key=self.encryption_key,
        )

    def _run(self, session_key: str, slot: str, data: Dict[str, Any]) -> str:
        async def _write() -> Any:
            async with self._client as client:
                await client.ephemeral.write_slot(session_key, slot, data)

        _run_sync(_write())
        return f'Wrote data to slot "{slot}" in session {session_key[:12]}...'

    async def _arun(self, session_key: str, slot: str, data: Dict[str, Any]) -> str:
        async with self._client as client:
            await client.ephemeral.write_slot(session_key, slot, data)
        return f'Wrote data to slot "{slot}" in session {session_key[:12]}...'


class XacheEphemeralReadSlotTool(BaseTool):
    """Read data from an ephemeral session slot."""

    name: str = "xache_ephemeral_read_slot"
    description: str = (
        "Read data from an ephemeral session slot. "
        "Slots: conversation, facts, tasks, cache, scratch, handoff."
    )
    args_schema: type[BaseModel] = EphemeralReadSlotInput

    wallet_address: str
    private_key: Optional[str] = Field(default=None, exclude=True)
    signer: Optional[Any] = Field(default=None, exclude=True)
    wallet_provider: Optional[Any] = Field(default=None, exclude=True)
    encryption_key: Optional[str] = Field(default=None, exclude=True)
    api_url: str = "https://api.xache.xyz"
    chain: str = "base"

    _client: Optional[XacheClient] = None

    def __init__(self, **kwargs: Any) -> None:
        import os

        if "api_url" not in kwargs or kwargs["api_url"] is None:
            kwargs["api_url"] = os.environ.get("XACHE_API_URL", "https://api.xache.xyz")
        super().__init__(**kwargs)
        chain_prefix = "sol" if self.chain == "solana" else "evm"
        did = f"did:agent:{chain_prefix}:{self.wallet_address.lower()}"
        self._client = XacheClient(
            api_url=self.api_url,
            did=did,
            private_key=self.private_key,
            signer=self.signer,
            wallet_provider=self.wallet_provider,
            encryption_key=self.encryption_key,
        )

    def _run(self, session_key: str, slot: str) -> str:
        import json

        async def _read() -> Any:
            async with self._client as client:
                return await client.ephemeral.read_slot(session_key, slot)

        data = _run_sync(_read())
        return json.dumps(data, indent=2)

    async def _arun(self, session_key: str, slot: str) -> str:
        import json

        async with self._client as client:
            data = await client.ephemeral.read_slot(session_key, slot)
        return json.dumps(data, indent=2)


class XacheEphemeralPromoteTool(BaseTool):
    """Promote an ephemeral session to persistent memory."""

    name: str = "xache_ephemeral_promote"
    description: str = (
        "Promote an ephemeral session to persistent memory. "
        "Extracts valuable data from slots and stores as permanent memories."
    )
    args_schema: type[BaseModel] = EphemeralPromoteInput

    wallet_address: str
    private_key: Optional[str] = Field(default=None, exclude=True)
    signer: Optional[Any] = Field(default=None, exclude=True)
    wallet_provider: Optional[Any] = Field(default=None, exclude=True)
    encryption_key: Optional[str] = Field(default=None, exclude=True)
    api_url: str = "https://api.xache.xyz"
    chain: str = "base"

    _client: Optional[XacheClient] = None

    def __init__(self, **kwargs: Any) -> None:
        import os

        if "api_url" not in kwargs or kwargs["api_url"] is None:
            kwargs["api_url"] = os.environ.get("XACHE_API_URL", "https://api.xache.xyz")
        super().__init__(**kwargs)
        chain_prefix = "sol" if self.chain == "solana" else "evm"
        did = f"did:agent:{chain_prefix}:{self.wallet_address.lower()}"
        self._client = XacheClient(
            api_url=self.api_url,
            did=did,
            private_key=self.private_key,
            signer=self.signer,
            wallet_provider=self.wallet_provider,
            encryption_key=self.encryption_key,
        )

    def _run(self, session_key: str) -> str:
        async def _promote() -> Any:
            async with self._client as client:
                return await client.ephemeral.promote_session(session_key)

        result = _run_sync(_promote())
        output = f"Promoted session {session_key[:12]}...\n"
        output += f"Memories created: {result.memories_created}\n"
        if result.memory_ids:
            output += f"Memory IDs: {', '.join(result.memory_ids)}\n"
        if result.receipt_id:
            output += f"Receipt: {result.receipt_id}"
        return output

    async def _arun(self, session_key: str) -> str:
        async with self._client as client:
            result = await client.ephemeral.promote_session(session_key)
        output = f"Promoted session {session_key[:12]}...\n"
        output += f"Memories created: {result.memories_created}\n"
        if result.memory_ids:
            output += f"Memory IDs: {', '.join(result.memory_ids)}\n"
        if result.receipt_id:
            output += f"Receipt: {result.receipt_id}"
        return output


class XacheEphemeralStatusTool(BaseTool):
    """Get ephemeral session status and details."""

    name: str = "xache_ephemeral_status"
    description: str = (
        "Get the status and details of an ephemeral session. "
        "Shows active slots, size, TTL, and window information."
    )
    args_schema: type[BaseModel] = EphemeralStatusInput

    wallet_address: str
    private_key: Optional[str] = Field(default=None, exclude=True)
    signer: Optional[Any] = Field(default=None, exclude=True)
    wallet_provider: Optional[Any] = Field(default=None, exclude=True)
    encryption_key: Optional[str] = Field(default=None, exclude=True)
    api_url: str = "https://api.xache.xyz"
    chain: str = "base"

    _client: Optional[XacheClient] = None

    def __init__(self, **kwargs: Any) -> None:
        import os

        if "api_url" not in kwargs or kwargs["api_url"] is None:
            kwargs["api_url"] = os.environ.get("XACHE_API_URL", "https://api.xache.xyz")
        super().__init__(**kwargs)
        chain_prefix = "sol" if self.chain == "solana" else "evm"
        did = f"did:agent:{chain_prefix}:{self.wallet_address.lower()}"
        self._client = XacheClient(
            api_url=self.api_url,
            did=did,
            private_key=self.private_key,
            signer=self.signer,
            wallet_provider=self.wallet_provider,
            encryption_key=self.encryption_key,
        )

    def _run(self, session_key: str) -> str:
        async def _status() -> Any:
            async with self._client as client:
                return await client.ephemeral.get_session(session_key)

        session = _run_sync(_status())
        if not session:
            return f"Session {session_key[:12]}... not found."

        slots = ", ".join(session.active_slots) if session.active_slots else "none"
        return (
            f"Session: {session.session_key[:12]}...\n"
            f"Status: {session.status}\n"
            f"Window: {session.window}/{session.max_windows}\n"
            f"TTL: {session.ttl_seconds}s\n"
            f"Expires: {session.expires_at}\n"
            f"Active Slots: {slots}\n"
            f"Total Size: {session.total_size} bytes\n"
            f"Cumulative Cost: ${session.cumulative_cost:.4f}"
        )

    async def _arun(self, session_key: str) -> str:
        async with self._client as client:
            session = await client.ephemeral.get_session(session_key)
        if not session:
            return f"Session {session_key[:12]}... not found."

        slots = ", ".join(session.active_slots) if session.active_slots else "none"
        return (
            f"Session: {session.session_key[:12]}...\n"
            f"Status: {session.status}\n"
            f"Window: {session.window}/{session.max_windows}\n"
            f"TTL: {session.ttl_seconds}s\n"
            f"Expires: {session.expires_at}\n"
            f"Active Slots: {slots}\n"
            f"Total Size: {session.total_size} bytes\n"
            f"Cumulative Cost: ${session.cumulative_cost:.4f}"
        )
